### ISSUE TITLE
Spatial Automation (across step)

### ISSUE THREAD

**Issue by matthewpereira** [concourse/concourse#5656]
### TLDR

This proposal introduces a mechanism by which a build plan can be executed multiple times with different ((vars)), in service of build matrices, pipeline matrices, and - building on ((var)) sources - dynamic variations of both.

The across step is given a set of values, a name for a ((var)), and a sub-plan to execute with each value assigned as the named var.

### Problem

* Modern workflow automation and application development requires tasks be run several times with different conditions - operating systems, dependency versions, etc. - and copy-pasting blocks of code to achieve this can cause mistakes that are difficult to spot.
* Concourse's YAML configuration can be daunting to beginners, and large and complicated pipeline.yml files can be daunting to edit. Anything we can do to DRY out our YAML is a benefit to devs and operators.
* In the web UI, it can be challenging to recognize and navigate groups of like tasks to make intelligent conclusions about task success and failure when you've manually configured a matrix of builds

### Proposed Solution

- Create a feature in Concourse that allows the user to provide a task with arrays of `((vars))`, and automatically create/run a task for each resulting configuration.

- This is related to the idea of instance vars used in the #5664 - Pipeline Instances feature. Instead of interpolating an n-dimensional group of pipelines using `n` lists of variables, this feature works at the build task level to interpolate an n-dimensional matrix of tasks from `n` lists of variables.

### Outcomes

- Creating a dynamic 'across' step will allow a job to monitor a set of vars, and trigger whenever that set changes. This allows pipeline automation on a new level than is currently possible.
- Concourse can perform automatic pipeline generation and automatic pipeline archiving.

### Current Status

- Engineering work started in #5887 - atc/web: static across steps

### References

[RFC conversation](https://github.com/concourse/rfcs/pull/29)
[Rendered RFC](https://github.com/vito/rfcs/blob/spatial-resources/029-across-step/proposal.md)

**Comment by aoldershaw** [comment]
@scottietremendous just FYI I personally don't consider spatial automation/the across step "done" - there's still the dynamic `across` step of course, but also, the static `across` step doesn't end up being entirely useful just yet due to limitations with var interpolation - one report is https://github.com/concourse/concourse/discussions/6061, and we encountered the same thing when we tried to apply the `across` step on our pipeline (and had to hack our way around it). I've been working on improving var interpolation on freedom fridays, but it's still a WIP.

**Comment by scottietremendous** [comment]
@aoldershaw - Good call. I'll link out to those issues. Maybe create a new epic for them.

**Comment by piroux** [comment]
Hi,

:arrow_right:  This is an important feature of the v10 Concourse roadmap, however it seems a bit blocked:
- in the [discussion of the specification](https://github.com/concourse/rfcs/pull/29)
- in the [PR of the experimental implementation](https://github.com/concourse/concourse/pull/5887)

:information_source:  Indeed, as most Concourse users must know by now, the stake is that we won't be able to properly and cleanly generate and handle PR builds, until this is done. (I know we can still do it in a way, but this is far from perfect without dynamic `across`).

:question: **How to move this forward ?**
:question: **How can community members can help out the Concourse team on this specific `across` feature ?**

:dart:  How about a list of action steps to move closer toward the specification and implementation of the dynamic `across` ?
:dart:  Or is it not possible to implement it the way this was initially envisioned ?

**Comment by aoldershaw** [comment]
Hey @piroux, thanks for bringing this up! It's still very important to us to support these workflows, but unfortunately we haven't been able to focus all of our efforts on it. We have been slowly chipping away at the v10 roadmap, but there are several things that need to be implemented before we can start supporting PR workflows:

1. The dynamic `across` step itself. As you note, the RFC is still open, but we feel we have a fairly good sense of how it'll work. I'll touch on this later in this comment.
2. Prototypes. We've recently implemented the "skeleton" for prototypes in #7065 (mostly boilerplate to be able to define prototypes in the pipeline config). However, we still need the following:
    1. #7061 - this is mainly about actually implementing the protocol itself and providing a way to run prototypes in pipelines.
        * There's still the open question about how to provide inputs and outputs to prototypes (see concourse/rfcs#103 - most of the discussion is in the comments).
        * I plan on writing up some notes on how to proceed with this issue in the next few days
    2. #6601 - this change allows prototypes to replace resource types (i.e. the `type` of a `resource` can be a `prototype`). This involves implementing a prototype "interface" (i.e. set of messages that it responds to, for `check`, `get`, and optionally `put`, `delete`) It also involves adding a new `delete` step to the pipeline config, as outlined in https://github.com/concourse/rfcs/blob/master/038-resource-prototypes/proposal.md
    3. #6275 - this also involves another prototype "interface". However, this one isn't quite as fleshed out as resource prototypes. Var source prototypes will likely respond to (at least) two messages: `list` and `get`.
        * A `github-pr` prototype, for instance, would implement at least the "var source" interface (to be able to list PRs), and may also implement the "resource" interface to track new commits to the PR (though this is probably better suited for the `git` resource...)

As you can tell, we still have a fair bit of work to do, and **we would love any contributions from the community on any of these features**. If you (or anyone coming across this thread) is interesting in contributing here, let us know and we'll try to provide more information/give suggestions if you get stuck.

---

Now, back to the dynamic `across` step. To start out with, I think it makes sense to simply allow the `across` step to interpolate `values`, since this gets you pretty far:

```yaml
across:
- var: foo
  values: [1, 2, 3] # static list
- var: bar
  values: ((.:some_var)) # dynamic list
```

Implementation-wise, we'd do the following:
1. Interpolate the values
2. Construct the plans for the sub-steps
3. Emit a new type of build event containing the plans for the sub-steps and the values of the interpolated variables, similar to how we dynamically send the "fetch image" plan as of Concourse 7.0:

https://github.com/concourse/concourse/blob/459b727875456577dbe94422528988f9dc62f790/atc/event/events.go#L226-L233

The UI would use this build event to the `across` step properly.

4. Run the generated plans, just like how we dynamically run the "fetch image" plan:

https://github.com/concourse/concourse/blob/459b727875456577dbe94422528988f9dc62f790/atc/engine/build_step_delegate.go#L278-L281

Happy to provide more information/context here too.

---

I realize that with just a dynamic `across` step, we could achieve PR workflows **without any of the prototype work**, though the results aren't quite as nice. There could be a e.g. `github-prs` resource type that produces a file like:

`prs.json`
```json
[
  {"number": 123, "uri": "git@github.com/..."},
  {"number": 456, "uri": "git@github.com/..."}
]
```

Then, we could have a job like:

```yaml
- get: ci
- get: repo-prs
  trigger: true
- load_var: prs
  file: repo-prs/prs.json
- across:
  - var: pr
    values: ((.:prs))
  set_pipeline: pr
  file: ci/pipelines/pr.yml
  instance_vars: {number: ((.:pr.number))}
  vars: {uri: ((.:pr.uri))}
```

**Comment by aoldershaw** [comment]
@piroux I got inspired by this and figured I might as well implement what I described around the dynamic across step - so I implemented #7252. Hopefully we can get it in for the next release, since I think it'll several workflows even without prototypes.

**Comment by aoldershaw** [comment]
I forked [telia-oss/github-pr-resource](https://github.com/telia-oss/github-pr-resource) to create [aoldershaw/github-prs-resource](https://github.com/aoldershaw/github-prs-resource) as a proof of concept for PR workflows using dynamic `across`. Unlike the telia-oss resource, this resource emits a list of PRs as a single version - the `get` step simply produces a file containing the list of PR numbers. A new version is emitted only when the set of PRs changes.

Testing it out with the following pipeline:

```yaml
resource_types:
- name: github-prs
  type: registry-image
  source:
    repository: aoldershaw/github-prs-resource

resources:
- name: prs
  type: github-prs
  source:
    repository: concourse/concourse
    access_token: ((github_access_token))

- name: mock
  type: mock
  source:
    create_files:
      pipeline.yml: |
        ...

jobs:
- name: across
  plan:
  - get: prs
    trigger: true
  - get: mock
  - load_var: pr_numbers
    file: prs/prs.json
  - across:
    - var: pr_number
      values: ((.:pr_numbers))
    set_pipeline: pr
    file: mock/pipeline.yml
    instance_vars: {number: ((.:pr_number))}
```

...and it's able to dynamically grow/shrink the `pr` instance group, thanks to archiving.

<img width="297" alt="Screen Shot 2021-07-07 at 9 15 06 AM" src="https://user-images.githubusercontent.com/26583442/124765314-d49eb980-df03-11eb-96d4-11f0e0b4b5b7.png">
